peer/peerdiscovery.go

```diff
@@ -0,0 +1,174 @@
+package peer
+
+import (
+	"context"
+	"encoding/binary"
+	"fmt"
+	"time"
+
+	"github.com/renproject/aw/transport"
+	"github.com/renproject/aw/wire"
+	"github.com/renproject/id"
+
+	"go.uber.org/zap"
+)
+
+type SignatoryAndAddress struct {
+	Signatory id.Signatory
+	Address   wire.Address
+}
+
+func (sigAndAddr SignatoryAndAddress) Marshal(buf []byte, rem int) ([]byte, int, error) {
+	buf, rem, err := sigAndAddr.Signatory.Marshal(buf, rem)
+	if err != nil {
+		return buf, rem, fmt.Errorf("marshal signatory: %v", err)
+	}
+	buf, rem, err = sigAndAddr.Address.Marshal(buf, rem)
+	if err != nil {
+		return buf, rem, fmt.Errorf("marshal address: %v", err)
+	}
+	return buf, rem, err
+}
+func (sigAndAddr *SignatoryAndAddress) Unmarshal(buf []byte, rem int) ([]byte, int, error) {
+	buf, rem, err := (&sigAndAddr.Signatory).Unmarshal(buf, rem)
+	if err != nil {
+		return buf, rem, fmt.Errorf("unmarshal signatory: %v", err)
+	}
+	buf, rem, err = sigAndAddr.Address.Unmarshal(buf, rem)
+	if err != nil {
+		return buf, rem, fmt.Errorf("unmarshal address: %v", err)
+	}
+	return buf, rem, err
+}
+
+type DiscoveryClient struct {
+	opts DiscoveryOptions
+
+	transport *transport.Transport
+}
+
+func NewDiscoveryClient(opts DiscoveryOptions, transport *transport.Transport) *DiscoveryClient {
+	return &DiscoveryClient{
+		opts: opts,
+
+		transport: transport,
+	}
+}
+
+func (dc *DiscoveryClient) PeerDiscovery(ctx context.Context) {
+	var pingData [2]byte
+	binary.LittleEndian.PutUint16(pingData[:], dc.transport.Port())
+
+	msg := wire.Msg{
+		Version: wire.MsgVersion1,
+		Type:    wire.MsgTypePing,
+		Data:    pingData[:],
+	}
+
+	ticker := time.NewTicker(dc.opts.PingTimePeriod)
+	for {
+		for _, sig := range dc.transport.Table().Peers(dc.opts.Alpha) {
+			msg.To = id.Hash(sig)
+			err := dc.transport.Send(ctx, sig, msg)
+			if err != nil {
+				dc.opts.Logger.Debug("pinging", zap.Error(err))
+			}
+		}
+		select {
+		case <-ctx.Done():
+			return
+		case <-ticker.C:
+		}
+	}
+}
+
+func (dc *DiscoveryClient) DidReceiveMessage(from id.Signatory, msg wire.Msg) error {
+	switch msg.Type {
+	case wire.MsgTypePing:
+		if err := dc.didReceivePing(from, msg); err != nil {
+			return err
+		}
+	case wire.MsgTypePingAck:
+		if err := dc.didReceivePingAck(from, msg); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (dc *DiscoveryClient) didReceivePing(from id.Signatory, msg wire.Msg) error {
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+
+	var addressBuf [1024]byte
+	addressByteSlice := addressBuf[:0]
+	port := binary.LittleEndian.Uint16(msg.Data)
+	ipAddr, ipAddrOk := dc.transport.Table().IP(from)
+	if !ipAddrOk {
+		return fmt.Errorf("ip address for remote peer not found")
```

Comment:
Instead of returning an error immediately, you should check if the peer already has a known `wire.Address` in the table. If so, then return nil, instead of returning an error.